### ISSUE TITLE
multiple-line environment variable (feature request for web)

### ISSUE THREAD

**Issue by KSXGitHub** [travis-ci/travis-ci#7715]
Setup multiple environment variables is a pain so I wish I could write/paste a single text contains everything (e.g. JSON, YAML, scripts, etc.)

**Comment by BanzaiMan** [comment]
Do you want to set

```bash
EVERYTHING="JSON_DATA=x
YAML_DATA=y
SCRIPT=z"
```

?

Or do you have a more concrete use case?

**Comment by KSXGitHub** [comment]
@BanzaiMan I mean

```bash
EVERYTHING_IN_JSON='{
   "foo": ["abc", "def"],
   "bar": ["ghi", "jkl"],
   "token": {"github": "..."}
}'

EVERYTHING_IN_YAML="
  foo:
    - abc
    - def
  bar:
    - jkl
    - ghi
  token:
    github: '...'
"

EVERYTHING_IN_SCRIPT="function () {
  // JavaScript
  const foo = ['abc', 'def']
  const bar = ['ghi', 'jkl']
  const token = {github: '...'}
  return {foo, bar}
}"
```

Then, in my script, I can just parse one of these "EVERYTHING" to get all I want

In addition, I also wish that travis could parse `EVERYTHING_IN_JSON` or `EVERYTHING_IN_YAML` for me to get `token.github` to pass to .travis.yml's `deploy.api_key`

**Comment by BanzaiMan** [comment]
I am not sure how practical this is. The need seems to me quite niche; if you need complex variables such as these, you would be better off writing them out to a file and `source`ing it.

**Comment by KSXGitHub** [comment]
@BanzaiMan 

I could, if only:

* These variables are no secret. (Think about a bunch of tokens for an open source project)
* These variables are the same across forks.

**Comment by GabrielDelepine** [comment]
I found a useful trick to export multiline variable 

1.replace all EOL (line break) by `\n`
2. In the web interface, copy-paste the value with the following prefix `"$(echo -e '` and suffix `')"`

For example, to export 
```
-----BEGIN CERTIFICATE-----
MKOQ1zYTmd
zkRpTFq4xu
-----BEGIN RSA PRIVATE KEY-----
```

replace EOL with `\n` and escape special characters such as space
```
-----BEGIN\ CERTIFICATE-----\nMKOQ1zYTmd\nzkRpTFq4xu\n-----BEGIN\ RSA PRIVATE\ KEY-----
```

add prefix and suffix
`"$(echo -e '` `-----BEGIN\ CERTIFICATE-----\nMKOQ1zYTmd\nzkRpTFq4xu\n-----BEGIN\ RSA PRIVATE\ KEY-----` `')"`

copy-paste it to the Travis interface
![selection_311](https://user-images.githubusercontent.com/1979087/35726622-ce91727c-0805-11e8-84bf-c30322a21a9f.png)

**Comment by KSXGitHub** [comment]
@GabrielDelepine Don't you see how hard it is to read?

**Comment by GabrielDelepine** [comment]
@KSXGitHub I agree with you. It's just a temporary solution for every people who are looking for a trick immediately available

**Comment by GabLeRoux** [comment]
Thanks, I made [this travis example project here](https://github.com/GabLeRoux/travis-multiline-env-example) to confirm that [GabrielDelepine's solution](https://github.com/travis-ci/travis-ci/issues/7715#issuecomment-362536708) works. 👍

**Comment by KSXGitHub** [comment]
@GabLeRoux I think you missed the point.

The problem is **not** "it is _impossible_ to set multi-line variable" **but** "setting multi-line variable is _inconvenient_ and _hard to read_"

**Comment by GabLeRoux** [comment]
Sure I know, google led me here, I confirm it's a pain, but at least it works. I wish it was easier. In fact, it works better on gitlab-ci, you can paste multi-line env vars directly and it will be left intact when using `echo "$SOME_ENV_VAR"` (quotes are important tho). :)

**Comment by bscubed** [comment]
I think the solution here would be to make the variable form on the Travis settings webpage a multi-line form instead of just a single-line form that would allow users to add multi-line variables wrapped with single quotes.

Because I can promise you that multi-line variables aren't just some niche thing, I've had to use the ugly workaround @GabrielDelepine mentioned (no offense) for things like SSH and GPG keys I didn't feel comfortable including on GitHub. You say multi-line variables are hard to read but wrapping everything in `"$(echo -e '` and `')"` is even harder to read.

**Comment by KSXGitHub** [comment]
@bscubed Finally, someone gets me :smile: 

The simplest solution is just to use `<textarea>`, how hard could it be!

**Comment by bscubed** [comment]
Also another (slightly less ugly) workaround is to wrap the variable with `$'` on the left side and `'` on the right side. You still have to escape new lines with `/n` though.

So you can save a variable called `SSH_KEY` in Travis web settings with the value `$'-----BEGIN CERTIFICATE-----\nMKOQ1zYTmd\nzkRpTFq4xu\n-----END CERTIFICATE-----'`, and the output of `echo "$SSH_KEY"` would be
```
-----BEGIN CERTIFICATE-----
MKOQ1zYTmd
zkRpTFq4xu
-----END CERTIFICATE-----
```

**Comment by msm1089** [comment]
> I found a useful trick to export multiline variable
> 
> 1.replace all EOL (line break) by `\n`
> 2. In the web interface, copy-paste the value with the following prefix `"$(echo -e '` and suffix `')"`
> 
> For example, to export
> 
> ```
> -----BEGIN CERTIFICATE-----
> MKOQ1zYTmd
> zkRpTFq4xu
> -----END CERTIFICATE-----
> ```
> 
> replace EOL with `\n` ~and escape special characters such as space~ edit : you don't need to escape other characters
> 
> ```
> -----BEGIN CERTIFICATE-----\nMKOQ1zYTmd\nzkRpTFq4xu\n-----END CERTIFICATE-----
> ```
> 
> add prefix and suffix
> `"$(echo -e '` `-----BEGIN CERTIFICATE-----\nMKOQ1zYTmd\nzkRpTFq4xu\n-----END CERTIFICATE-----` `')"`
> 
> copy-paste it to the Travis interface
> ![selection_311](https://user-images.githubusercontent.com/1979087/35726622-ce91727c-0805-11e8-84bf-c30322a21a9f.png)

Thanks for this trick - it finally allowed me to get a test to to pass.

I was trying to add an RSA key as an environment variable in the travis.yml...I tried everything I could find by Googling but nothing seemed to work. Anyone know how to add it to the .yml file?

**Comment by JanMikes** [comment]
👍 Thank you very much for this issue. After 2-3 hours trying to get my private key in variable work, i found this and finally wrapping it wih `$''` works.

This should either be better documented or implemented and supported multiline env vars out of the bos.

Me as pro user of Gitlab CI where it works by default, would expect the same in travis ci :-)

**Comment by oakbani** [comment]
@GabrielDelepine  Thanks a lot for your solution. Finally reached this after spending hours.

**Comment by KSXGitHub** [comment]
It's been 3 years and no one on the dev team notice this?

**Comment by GabLeRoux** [comment]
There's a workaround and changing the behavior may introduce breaking changes I guess. They could do something similar to what gitlab-ci did with its variables (they added file support):

![image](https://user-images.githubusercontent.com/1264761/72063214-5e39e080-32a7-11ea-8a94-e0523107264f.png)

https://docs.gitlab.com/ee/ci/variables/#file-type

**Comment by kvlaznev** [comment]
In my Laravel project i'm using this workaround:

&& echo "$ENV_CONFIG_DEPLOY" | tr ' ' '\n' > /app/$CI_PROJECT_NAME/.env

Environment ENV_CONFIG_DEPLOY in CI Setting -> Variables like:

APP_DEBUG=false
APP_LOG=stack
LOG_CHANNEL=stack
APP_LOG_LEVEL=info
APP_LOG_MAX_FILES=30
APP_LOG_FLUENTD_HOST=
APP_LOG_FLUENTD_PORT=24224
